### PR TITLE
[Framework] implement ingress operator

### PR DESCRIPTION
**Pull request by weilerN**
### **Description**  
Full description in the LLD that linked to this Jira - https://iguazio.atlassian.net/browse/NUC-498
This PR introduces a new IngressWatcher component that uses Kubernetes informers to watch for Ingress resource changes and updates an internal IngressHostCache accordingly. The watcher handles add, update, and delete events, extracts relevant data (host, path, target list), and reflects those changes in the in-memory cache.

### **Affected Areas**  
- `pkg/kube/ingress.go` – new file with the IngressWatcher implementation
- Uses `IngressHostCache` from `pkg/ingresscache`
- Integrates with `k8s.io/client-go` for informers

### **Testing**  
- Unit tests coverage for all event handler paths (add, update, delete) correctly modify the ingress cache.
- Unit tests validates cache sync behavior and logging of edge cases (e.g., missing host/path).
- Integrations tests are out of scope for this PR, will be introduced as part of the code changes in `Nuclio`

### **Changes Made**  
- Added `IngressWatcher` struct to manage and respond to Kubernetes Ingress resource changes.
- Implemented handlers for Add, Update, and Delete events.
- Extracted ingress data: host, path, targets via helper methods.
- Registered shared informer factory with label selector support.

### **Additional notes** 
- Assumes each Ingress contains exactly one rule and one path.
- Includes detailed error logging for resilience and observability.
- Additional PR for integrating the ingress cache and ingress watcher is needed

### CODE REVIEW

File: pkg/kube/ingress.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	if err = iw.ingressCache.Delete(host, path, targets); err != nil {
+		iw.logger.WarnWith("Delete ingress handler failure- failed delete from cache", "error", err, "object", obj)
+		return
+	}
+}
+
+// --- internal methods ---
+
+// extractIngressValuesFromIngressResource extracts the host, path, and targets from the ingress resource.
+func (iw *IngressWatcher) extractIngressValuesFromIngressResource(obj interface{}) (string, string, []string, error) {
```
Comment (TomerShor) [inline pkg/kube/ingress.go:161]: targets is not an ingress value really
```suggestion
func (iw *IngressWatcher) extractValuesFromIngressResource(obj interface{}) (string, string, []string, error) {
```

File: pkg/kube/ingress_test.go [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+	var err error
+
+	suite.logger, err = nucliozap.NewNuclioZapTest("test")
+	suite.Require().NoError(err)
+}
+
+// TestIngressWatcherTestSuite runs the test suite
+func TestIngressWatcherTestSuite(t *testing.T) {
+	suite.Run(t, new(IngressWatcherTestSuite))
+}
```
Comment (TomerShor) [inline pkg/kube/ingress_test.go:58]: Move to the bottom of the file

File: pkg/kube/ingress_test.go [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+	type ingressArgs struct {
+		host   string
+		path   string
+		target []string
+	}
+	for _, testCase := range []struct {
+		name              string
+		testArgs          ingressArgs
+		expectedResult    []string
+		initialStateCache *ingressArgs
```
Comment (TomerShor) [inline pkg/kube/ingress_test.go:70]:
```suggestion
		initialCachedData *ingressArgs
```

File: pkg/kube/ingress_test.go [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+		})
+	}
+}
+
+func (suite *IngressWatcherTestSuite) TestIngressHandlerUpdateFunc() {
+	type ingressArgs struct {
+		host    string
+		path    string
+		targets []string
+	}
```
Comment (TomerShor) [inline pkg/kube/ingress_test.go:151]: Following @rokatyy 's suggestion, this can be a struct in `ingress.go`.
But generally - if you see that you defined the same struct multiple times - define it once outside of the tests.

File: pkg/kube/ingress_test.go [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+		host           string
+		path           string
+		targets        []string
+		expectError    bool
+		expectErrorMsg string
+	}
+	for _, testCase := range []struct {
+		name              string
+		expectedResults   []expectedResults
+		initialStateCache *ingressArgs
```
Comment (TomerShor) [inline pkg/kube/ingress_test.go:162]:
```suggestion
		initialCacheData *ingressArgs
```

File: pkg/kube/ingress_test.go [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+		path    string
+		targets []string
+	}
+	type expectedResult struct {
+		host           string
+		path           string
+		targets        []string
+		expectError    bool
+		expectErrorMsg string
+	}
```
Comment (TomerShor) [inline pkg/kube/ingress_test.go:292]: Same for this - move it to the top of the file.

**Review by TomerShor — CHANGES_REQUESTED** [review]

File: pkg/kube/ingress.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+	path, err := iw.getPathFromIngress(ingress)
+	if err != nil {
+		return "", "", nil, errors.Wrap(err, "Failed to extract path from ingress")
+	}
+
+	targets, err := iw.resolveTargetsCallback(ingress)
+	if err != nil {
+		return "", "", nil, errors.Wrap(err, "Failed to extract targets from ingress labels")
+	}
+
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go:191]: 1. I agree that since the callback function is external to the scaler, it's appropriate to check for nil before using the returned targets.
2. The order of the function calls isn't critical, but I originally structured them to follow the logical route hierarchy: `host -> path -> targets`. 
If I understood your preference correctly, you'd like the target resolution to come first since it’s the most complex part, and placing it earlier supports a fail-fast approach. That makes sense, and I’ll update the order accordingly.

File: pkg/kube/ingress.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+	// if the host or path has changed, we need to delete the old entry
+	if oldHost != newHost || oldPath != newPath {
+		if err = iw.ingressCache.Delete(oldHost, oldPath, oldTargets); err != nil {
+			iw.logger.WarnWith("Update ingress handler failure - failed to delete old ingress", "error", err)
+		}
+	}
+
+	if err = iw.ingressCache.Set(newHost, newPath, newTargets); err != nil {
+		iw.logger.WarnWith("Update ingress handler failure- failed to add the new value", "error", err, "object", newObj)
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go:140]: Thanks for raising this.

Regarding the concern about atomicity and potential race conditions:

1. **Handler execution is thread-safe** — the `SharedIndexInformer` processes events using a [DeltaFIFO queue](https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L543), which guarantees that `Add`, `Update`, and `Delete` handlers are executed sequentially. So there’s no concurrency between them.

2. **The cache itself is thread-safe** — `IngressCache` uses `sync.Map` for host-level operations and an internal RWMutex (via a safeTrie) for path-level operations. This ensures safety even when accessed concurrently from multiple goroutines.

As for the concern about potential race conditions caused by other mechanisms (like refreshing the cache on a miss) — that’s a broader concern but out of scope for this PR. This change only introduces the informer-based handlers, which operate in a controlled and sequential environment. 
If needed, system-level synchronization between external components and the cache can be reviewed separately.

File: pkg/kube/ingress_test.go (hunk 6 above)
Comment (weilerN) [reply pkg/kube/ingress_test.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress_test.go (hunk 4 above)
Comment (weilerN) [reply pkg/kube/ingress_test.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress_test.go (hunk 3 above)
Comment (weilerN) [reply pkg/kube/ingress_test.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress_test.go (hunk 2 above)
Comment (weilerN) [reply pkg/kube/ingress_test.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go (hunk 1 above)
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+// --- ResourceEventHandler methods ---
+
+func (iw *IngressWatcher) IngressHandlerAddFunc(obj interface{}) {
+	host, path, targets, err := iw.extractIngressValuesFromIngressResource(obj)
+	if err != nil {
+		iw.logger.WarnWith("Add ingress handler failure", "error", err)
+		return
+	}
+
+	if err = iw.ingressCache.Set(host, path, targets); err != nil {
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+		iw.logger.WarnWith("Add ingress handler failure- failed to add the new value", "error", err, "object", obj)
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+	if !cache.WaitForCacheSync(iw.ctx.Done(), iw.informer.HasSynced) {
+		return errors.New("failed to sync ingress cache")
+	}
+
+	return nil
+}
+
+func (iw *IngressWatcher) Stop() {
+	iw.logger.Debug("Stopping IngressWatcher")
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+func (iw *IngressWatcher) Start() error {
+	iw.logger.Debug("Starting IngressWatcher")
+	iw.factory.Start(iw.ctx.Done())
+
+	if !cache.WaitForCacheSync(iw.ctx.Done(), iw.informer.HasSynced) {
+		return errors.New("failed to sync ingress cache")
+	}
+
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go:128]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	ctx context.Context,
+	dlxLogger logger.Logger,
+	kubeClient kubernetes.Interface,
+	ingressCache *ingresscache.IngressCache,
+	resolveCallback ResolveTargetsFromIngressCallback,
+	namespace, labelsFilter string,
+) (*IngressWatcher, error) {
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		kubeClient,
+		30*time.Second,
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+		informers.WithNamespace(namespace),
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.LabelSelector = labelsFilter
+		}),
+	)
+	ingressInformer := factory.Networking().V1().Ingresses().Informer()
+
+	ingressWatcher := &IngressWatcher{
+		ctx:                    ctx,
+		logger:                 dlxLogger,
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+	newHost, newPath, newTargets, err := iw.extractIngressValuesFromIngressResource(newObj)
+	if err != nil {
+		iw.logger.WarnWith("Update ingress handler - failed to extract values from new object", "error", err)
+		return
+	}
+
+	// if the host or path has changed, we need to delete the old entry
+	if oldHost != newHost || oldPath != newPath {
+		if err = iw.ingressCache.Delete(oldHost, oldPath, oldTargets); err != nil {
+			iw.logger.WarnWith("Update ingress handler failure - failed to delete old ingress", "error", err)
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+	if rule.HTTP == nil {
+		return "", errors.New("no HTTP configuration found in ingress rule")
+	}
+
+	if len(rule.HTTP.Paths) == 0 {
+		return "", errors.New("no paths found in ingress HTTP rule")
+	}
+
+	// Ingress must contain exactly one rule
+	httpPath := rule.HTTP.Paths[0]
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+	host, path, targets, err := iw.extractIngressValuesFromIngressResource(obj)
+	if err != nil {
+		iw.logger.WarnWith("Delete ingress handler failure- failed to extract values from object", "error", err)
+		return
+	}
+
+	if err = iw.ingressCache.Delete(host, path, targets); err != nil {
+		iw.logger.WarnWith("Delete ingress handler failure- failed delete from cache", "error", err, "object", obj)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/fc972c82145e64d4022498dac09dbc3cabf6820d)

File: pkg/kube/ingress.go [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+		return nil, errors.New("Failed to cast object to Ingress")
+	}
+
+	targets, err := iw.resolveTargetsCallback(ingress)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to extract targets from ingress labels")
+	}
+
+	if len(targets) == 0 {
+		return nil, errors.New("No targets found in ingress")
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go:262]: Fixed here - [CR comments - handle canary ingress resource failure](https://github.com/v3io/scaler/pull/75/commits/d00d4c61c8c7e630905f34e4a8fda00144acc762)

[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go:232]: As Discussed F2F - this comment was reverted and log severity changes fixe here - [CR comments - change warning logs](https://github.com/v3io/scaler/pull/75/commits/05227815628428188a76bf8753f11f531d931a40)

File: pkg/kube/ingress.go [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	return ingressWatcher, nil
+}
+
+func (iw *IngressWatcher) Start() error {
+	iw.logger.Info("Starting ingress watcher")
+	iw.factory.Start(iw.ctx.Done())
+
+	if !cache.WaitForCacheSync(iw.ctx.Done(), iw.informer.HasSynced) {
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go:125]: `iw.ctx.Done()` provides a cancellation channel so `WaitForCacheSync` doesn't block indefinitely if the context is cancelled or times out.

File: pkg/kube/ingress.go [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+func (iw *IngressWatcher) DeleteHandler(obj interface{}) {
+	ingress, err := iw.extractValuesFromIngressResource(obj)
+	if err != nil {
+		iw.logger.DebugWith("Delete ingress handler failure- failed to extract values from object",
+			"error", err.Error())
+		return
+	}
+
+	if err := iw.cache.Delete(ingress.host, ingress.path, ingress.targets); err != nil {
+		iw.logger.WarnWith("Delete ingress handler failure- failed delete from cache",
```
Comment (TomerShor) [inline pkg/kube/ingress.go:224]:
```suggestion
		iw.logger.WarnWith("Delete ingress handler failure - failed delete from cache",
```

**Review by TomerShor — CHANGES_REQUESTED** [review]
Last minor comments from me

File: pkg/kube/ingress.go (hunk 20 above)
Comment (weilerN) [reply pkg/kube/ingress.go]: Fixed here - [CR comments](https://github.com/v3io/scaler/pull/75/commits/2f8b01a3acc402dce56c39a362e2c56c788f48eb)

File: pkg/kube/ingress.go [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+			"object", newObj,
+			"ingress name", newIngress.name,
+			"host", newIngress.host,
+			"path", newIngress.path,
+			"targets", newIngress.targets)
+		return
+	}
+
+	iw.logger.DebugWith("Update ingress handler - successfully updated ingress in cache",
+		"ingress name", newIngress.name,
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: [CR comments](https://github.com/v3io/scaler/pull/75/commits/2f8b01a3acc402dce56c39a362e2c56c788f48eb)

File: pkg/kube/ingress.go [hunk 22]
```diff
@@ -0,0 +1,338 @@
...
+			"object", obj,
+			"ingress name", ingress.name,
+			"host", ingress.host,
+			"path", ingress.path,
+			"targets", ingress.targets)
+		return
+	}
+
+	iw.logger.DebugWith("Add ingress handler - successfully added ingress to cache",
+		"ingress name", ingress.name,
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: [CR comments](https://github.com/v3io/scaler/pull/75/commits/2f8b01a3acc402dce56c39a362e2c56c788f48eb)

File: pkg/kube/ingress.go [hunk 23]
```diff
@@ -0,0 +1,338 @@
...
+		iw.logger.DebugWith("Add ingress handler failure - failed to extract values from ingress resource",
+			"error", err.Error())
+		return
+	}
+
+	if err := iw.cache.Set(ingress.host, ingress.path, ingress.targets); err != nil {
+		iw.logger.WarnWith("Add ingress handler failure - failed to add the new value to ingress cache",
+			"error", err.Error(),
+			"object", obj,
+			"ingress name", ingress.name,
```
[the remark this answers is not in the document]
Comment (weilerN) [reply pkg/kube/ingress.go]: [CR comments](https://github.com/v3io/scaler/pull/75/commits/2f8b01a3acc402dce56c39a362e2c56c788f48eb)

**Review by TomerShor — APPROVED** [review]
